File: components/credentials/src/main/java/org/cloudfoundry/credhub/entity/CertificateCredentialVersionData.java
package org.cloudfoundry.credhub.entity;

import java.time.Instant;
import java.util.Objects;

import javax.persistence.Column;
import javax.persistence.DiscriminatorValue;
import javax.persistence.Entity;
import javax.persistence.PrimaryKeyJoinColumn;
import javax.persistence.SecondaryTable;

import org.apache.commons.lang3.StringUtils;

import static org.cloudfoundry.credhub.entity.CertificateCredentialVersionData.CREDENTIAL_DATABASE_TYPE;

@Entity
@DiscriminatorValue(CREDENTIAL_DATABASE_TYPE)
@SecondaryTable(
  name = CertificateCredentialVersionData.TABLE_NAME,
  pkJoinColumns = @PrimaryKeyJoinColumn(name = "uuid", referencedColumnName = "uuid")
)
public class CertificateCredentialVersionData extends CredentialVersionData<CertificateCredentialVersionData> {

  public static final String CREDENTIAL_DATABASE_TYPE = "cert";
  public static final String CREDENTIAL_TYPE = "certificate";
  public static final String TABLE_NAME = "certificate_credential";

  @Column(table = CertificateCredentialVersionData.TABLE_NAME, length = 7000)
  private String ca;

  @Column(table = CertificateCredentialVersionData.TABLE_NAME, length = 7000)
  private String certificate;

  @Column(table = CertificateCredentialVersionData.TABLE_NAME)
  private String caName;

  @Column(table = CertificateCredentialVersionData.TABLE_NAME)
  private boolean transitional;

  @Column(table = CertificateCredentialVersionData.TABLE_NAME)
  private Instant expiryDate;

  @Column(table = CertificateCredentialVersionData.TABLE_NAME)
  private Boolean certificateAuthority;

  @Column(table = CertificateCredentialVersionData.TABLE_NAME)
  private Boolean selfSigned;

  @Column(table = CertificateCredentialVersionData.TABLE_NAME)
  private Boolean generated;

  @Column(table = CertificateCredentialVersionData.TABLE_NAME, length = 7000, columnDefinition = "TEXT")
  private String trustedCa;

  public CertificateCredentialVersionData() {
    super();
  }

  public CertificateCredentialVersionData(final String name) {
    super(name);
  }

  public String getName() {
    return super.getCredential().getName();
  }

  public String getCa() {
    return ca;
  }

  public void setCa(final String ca) {
    this.ca = ca;
  }

  public String getCertificate() {
    return certificate;
  }

  public void setCertificate(final String certificate) {
    this.certificate = certificate;
  }

  public String getTrustedCa() {
    return trustedCa;
  }

  public void setTrustedCa(final String trustedCa) {
    this.trustedCa = trustedCa;
  }

  public String getCaName() {
    return caName;
  }

  public void setCaName(final String caName) {
    this.caName = !StringUtils.isEmpty(caName) ? StringUtils.prependIfMissing(caName, "/") : caName;
  }

  public Instant getExpiryDate() {
    return expiryDate;
  }

  public void setExpiryDate(final Instant expiryDate) {
    this.expiryDate = expiryDate;
  }

  @Override
  public String getCredentialType() {
    return CREDENTIAL_TYPE;
  }

  public boolean isTransitional() {
    return transitional;
  }

  public void setTransitional(final boolean transitional) {
    this.transitional = transitional;
  }

  public boolean isSelfSigned() {
    if (selfSigned == null) {
      return false;
    }

    return selfSigned;
  }

  public void setSelfSigned(final boolean selfSigned) {
    this.selfSigned = selfSigned;
  }

  public boolean isCertificateAuthority() {
    if (certificateAuthority == null) {
      return false;
    }

    return certificateAuthority;
  }

  public void setCertificateAuthority(final boolean certificateAuthority) {
    this.certificateAuthority = certificateAuthority;
  }

  public Boolean getGenerated() {
    return generated;
  }

  public void setGenerated(final Boolean generated) {
    this.generated = generated;
  }

  @Override
  public boolean equals(final Object o) {
    if (this == o) {
      return true;
    }
    if (o == null || getClass() != o.getClass()) {
      return false;
    }
    final CertificateCredentialVersionData that = (CertificateCredentialVersionData) o;
    return transitional == that.transitional &&
      Objects.equals(certificateAuthority, that.certificateAuthority) &&
      Objects.equals(selfSigned, that.selfSigned) &&
      Objects.equals(ca, that.ca) &&
      Objects.equals(certificate, that.certificate) &&
      Objects.equals(caName, that.caName) &&
      Objects.equals(expiryDate, that.expiryDate);
  }

  @Override
  public int hashCode() {
    return Objects.hash(ca, certificate, caName, transitional, expiryDate, certificateAuthority, selfSigned);
  }

  @Override
  public String toString() {
    return "CertificateCredentialVersionData{" +
      "ca='" + ca + '\'' +
      ", certificate='" + certificate + '\'' +
      ", caName='" + caName + '\'' +
      ", transitional=" + transitional +
      ", expiryDate=" + expiryDate +
      ", certificateAuthority=" + certificateAuthority +
      ", selfSigned=" + selfSigned +
      '}';
  }
}

File: components/credentials/src/main/kotlin/org/cloudfoundry/credhub/entity/CertificateCredentialVersionData.kt
package org.cloudfoundry.credhub.entity

import org.apache.commons.lang3.StringUtils
import org.cloudfoundry.credhub.entities.CredentialVersionData
import org.cloudfoundry.credhub.entity.CertificateCredentialVersionData.Companion.CREDENTIAL_DATABASE_TYPE
import java.time.Instant
import java.util.Objects
import javax.persistence.Column
import javax.persistence.DiscriminatorValue
import javax.persistence.Entity
import javax.persistence.PrimaryKeyJoinColumn
import javax.persistence.SecondaryTable

@Entity
@DiscriminatorValue(CREDENTIAL_DATABASE_TYPE)
@SecondaryTable(name = CertificateCredentialVersionData.TABLE_NAME, pkJoinColumns = [PrimaryKeyJoinColumn(name = "uuid", referencedColumnName = "uuid")])
class CertificateCredentialVersionData : CredentialVersionData<CertificateCredentialVersionData> {

    @Column(table = CertificateCredentialVersionData.TABLE_NAME, length = 7000)
    var ca: String? = null

    @Column(table = CertificateCredentialVersionData.TABLE_NAME, length = 7000)
    var certificate: String? = null

    @Column(table = CertificateCredentialVersionData.TABLE_NAME)
    var caName: String? = null
        set(caName) {
            field = if (!StringUtils.isEmpty(caName)) StringUtils.prependIfMissing(caName, "/") else caName
        }

    @Column(table = CertificateCredentialVersionData.TABLE_NAME)
    var transitional: Boolean = false

    @Column(table = CertificateCredentialVersionData.TABLE_NAME)
    var expiryDate: Instant? = null

    @Column(table = CertificateCredentialVersionData.TABLE_NAME)
    private var certificateAuthority: Boolean? = null

    @Column(table = CertificateCredentialVersionData.TABLE_NAME)
    private var selfSigned: Boolean? = null

    @Column(table = CertificateCredentialVersionData.TABLE_NAME, name = "certificate_generated")
    var generated: Boolean? = null

    @Column(table = CertificateCredentialVersionData.TABLE_NAME, length = 7000, columnDefinition = "TEXT")
    var trustedCa: String? = null

    val name: String?
        get() = super.credential!!.name

    override val credentialType: String
        get() = CREDENTIAL_TYPE

    var isSelfSigned: Boolean
        get() = if (selfSigned == null) {
            false
        } else selfSigned!!
        set(selfSigned) {
            this.selfSigned = selfSigned
        }

    var isCertificateAuthority: Boolean
        get() = if (certificateAuthority == null) {
            false
        } else certificateAuthority!!
        set(certificateAuthority) {
            this.certificateAuthority = certificateAuthority
        }

    // Needed for hibernate
    constructor() : super() {}

    constructor(name: String) : super(name) {}

    override fun equals(other: Any?): Boolean {
        if (this === other) {
            return true
        }
        if (other == null || javaClass != other.javaClass) {
            return false
        }
        val that = other as CertificateCredentialVersionData?
        return isTransitional() == that!!.isTransitional() &&
            certificateAuthority == that.certificateAuthority &&
            selfSigned == that.selfSigned &&
            ca == that.ca &&
            certificate == that.certificate &&
            this.caName == that.caName &&
            expiryDate == that.expiryDate
    }

    fun isTransitional(): Boolean {
        return this.transitional
    }

    override fun hashCode(): Int {
        return Objects.hash(ca, certificate, this.caName, isTransitional(), expiryDate, certificateAuthority, selfSigned)
    }

    override fun toString(): String {
        return "CertificateCredentialVersionData{" +
            "ca='" + ca + '\''.toString() +
            ", certificate='" + certificate + '\''.toString() +
            ", caName='" + this.caName + '\''.toString() +
            ", transitional=" + isTransitional() +
            ", expiryDate=" + expiryDate +
            ", certificateAuthority=" + certificateAuthority +
            ", selfSigned=" + selfSigned +
            '}'.toString()
    }

    companion object {
        const val CREDENTIAL_DATABASE_TYPE = "cert"
        const val CREDENTIAL_TYPE = "certificate"
        const val TABLE_NAME = "certificate_credential"
    }
}
